xmppserver/src/main/java/org/jivesoftware/openfire/http/HttpSession.java

```diff
@@ -909,41 +926,79 @@ private void deliver(@Nonnull final Deliverable stanza) {
+     * available for delivery.
+     *
+     * When no data is queued for delivery, or when no connection is available, an invocation does nothing.
+     */
+    private void tryImmediateDelivery()
+    {
+        final List<Deliverable> deliverables = drainPendingElements();
+        if (deliverables.isEmpty()) {
+            Log.trace("Immediate delivery of pending data to the client on session {} was requested, but no data is pending.", getStreamID());
+            return;
+        }
+
+        final HttpConnection connection = getConnectionReadyForOutboundDelivery();
+
+        if (connection == null) {
+            Log.trace("Immediate delivery of pending data to the client on session {} was requested, but no connection is available. The data ({} deliverables) will be re-queued.", getStreamID(), deliverables.size());
+            // place pending deliverables back on queue. // FIXME: if other threads have placed pending elements, this will cause a re-order, which might be undesirable.
+            pendingElements.addAll(deliverables);
+            return;
+        }
+
+        // OF-2444: deliver asynchronously, to avoid deadlocking issues.
+        HttpBindManager.getInstance().getSessionManager().execute(() -> {
+            try {
+                deliver(connection, deliverables, true);
+            } catch (HttpConnectionClosedException e) {
+                /* Connection was closed, try the next one. Indicates a (concurrency?) bug. */
+                Log.warn("Iterating over a connection that was closed for session {}. Openfire will recover from this problem, but it should not occur in the first place.", getStreamID(), e);
+                // place pending deliverables back on queue. // FIXME: if other threads have placed pending elements, this will cause a re-order, which might be undesirable.
+                pendingElements.addAll(deliverables);
+            } catch (IOException e) {
+                Log.warn("An unexpected exception occurred while iterating over connections for session {}. Openfire will attempt to recover by ignoring this connection.", getStreamID(), e);
+                // place pending deliverables back on queue. // FIXME: if other threads have placed pending elements, this will cause a re-order, which might be undesirable.
+                pendingElements.addAll(deliverables);
             }
+        });
+    }
 
-            if (!delivered) {
-                Log.debug("Unable to immediately deliver a stanza on session {}. It is being queued instead).", getStreamID());
-                pendingElements.addAll( deliverable );
+    /**
+     * Returns a connection that can be used to deliver data to the client, if such a connection is currently available.
+     *
+     * @return A connection that is ready to be responded to.
+     */
+    @Nullable
+    private HttpConnection getConnectionReadyForOutboundDelivery()
```

Comment:
Might be tempted to use an Optional here, to force handling of the 'not present' case

Reply:
I never learn to use those. Makes more sense than passing around null values. :)

I have added a commit to this PR that makes the change (found one more method where it applied).